src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -476,92 +245,223 @@ HTTPProxy::_handleCreateSchema(sh::dto::CreateSchemaRequest&& request) {
+seastar::future<std::tuple<sh::Status, shd::WriteResponse>>
+HTTPProxy::_handleWrite(shd::WriteRequest&& request) {
+    K2LOG_D(log::httpproxy, "Received write request {}", request);
+
+    return seastar::do_with(std::move(request), [this] (auto& request) {
+        return _getSchemas(request.collectionName, request.schemaName, request.value.schemaVersion)
+        .then([this, &request](auto&& schemas) mutable {
+            auto& [status, k2Schema, shdSchema] = schemas;
+            if (!status.is2xxOK()) {
+                return MakeHTTPResponse<shd::WriteResponse>(sh::Status{.code=status.code, .message=status.message}, shd::WriteResponse{});
+            }
+            auto it = _txns.find(request.timestamp);
+            if (it == _txns.end()) {
+                return MakeHTTPResponse<shd::WriteResponse>(sh::Statuses::S410_Gone("transaction does not exist"), shd::WriteResponse{});
+            }
+            dto::SKVRecord k2record(request.collectionName, k2Schema);
+            bool haveKeys = !request.isDelete && request.fieldsForPartialUpdate.size() == 0;
+            shd::SKVRecord shdrecord(request.schemaName, shdSchema, std::move(request.value), haveKeys);
+            _shdRecToK2(shdrecord, k2record);
+
+            return it->second.handle.write(k2record, request.isDelete, static_cast<dto::ExistencePrecondition>(request.precondition))
+                .then([](WriteResult&& result) {
+                    return MakeHTTPResponse<shd::WriteResponse>(sh::Status{.code = result.status.code, .message = result.status.message}, shd::WriteResponse{});
+                });
+        });
+    });
 }
 
-seastar::future<> HTTPProxy::gracefulStop() {
-    _stopped = true;
+seastar::future<std::tuple<sh::Status, shd::ReadResponse>>
+HTTPProxy::_handleRead(shd::ReadRequest&& request) {
+    K2LOG_D(log::httpproxy, "Received read request {}", request);
+
+    return seastar::do_with(std::move(request), [this](auto& request) {
+        return _getSchemas(request.collectionName, request.schemaName, request.key.schemaVersion)
+            .then([this, &request](auto&& schemas) mutable {
+                auto& [status, k2Schema, shdSchema] = schemas;
+                if (!status.is2xxOK()) {
+                    return MakeHTTPResponse<shd::ReadResponse>(sh::Status{.code = status.code, .message = status.message}, shd::ReadResponse{});
+                }
+                auto it = _txns.find(request.timestamp);
+                if (it == _txns.end()) {
+                    return MakeHTTPResponse<shd::ReadResponse>(sh::Statuses::S410_Gone("transaction does not exist"), shd::ReadResponse{});
+                }
+                dto::SKVRecord k2record(request.collectionName, k2Schema);
+                shd::SKVRecord shdrecord(request.schemaName, shdSchema, std::move(request.key), true);
+                _shdRecToK2(shdrecord, k2record);
+
+                return it->second.handle.read(k2record.getKey(), request.collectionName)
```

Comment:
Should we use the SKVRecord read instead?

Reply:
Yup, makes sense. no idea why i didn't do that in the first place :)